Bug 1836337: Azure: Add functionality to change Azure Machine Disk Types

rna-afk opened the pull request:
Azure disk types for the master and worker nodes is by default
set to Premium SSD(Premium_LRS) setting. Made the changes to add
the disktype of the master and worker nodes in the install config
by adding an extra field in MachinePool DiskType to the OSDisk struct
which is part of the DefaultMachinePlatform attribute in the Azure
Platform type. To change the disk type, the install config must be
changed by adding the diskType under platform->Azure property by adding
the defaultMachinePlatform->osDisk->diskType and setting the value to
one of the following, Premium_LRS, Standard_LRS and StandardSSD_LRS.

Setting Standard_LRS is not recommended for the master nodes and hence,
if the disktype is set to it then the Master node is defaulted to Premium_LRS
and the worker nodes will take up the Standard_LRS.

Also, DefaultMachinePlatform can only have both Premium_LRS or 
StandardSSD_LRS for the above reason.

Currently, the UltraSSD_LRS is under construction as the terraform version
needs to be updated.

patrickdillon commented:
/test e2e-azure

patrickdillon commented:
This PR should update docs as well.

rna-afk commented:
/test e2e-azure

rna-afk commented:
/test images

patrickdillon commented on pkg/types/azure/machinepool.go:
```diff
@@ -38,4 +44,18 @@ func (a *MachinePool) Set(required *MachinePool) {
+
+	a.OSDisk.DiskType = validateDiskType(required.OSDisk.DiskType)
+}
+func validateDiskType(diskType string) string {
```
validating and setting defaults are two different functions

rna-afk replied:
Couldn't find an appropriate function name for this. It validates the value and sets the default when there is empty or invalid value.

patrickdillon commented on pkg/types/azure/machinepool.go:
```diff
@@ -38,4 +44,18 @@ func (a *MachinePool) Set(required *MachinePool) {
+	if diskType == "" {
+		return "Premium_LRS"
+	}
+	_, found := map[string]string{"Standard_LRS": "", "StandardSSD_LRS": "", "Premium_LRS": ""}[ //"UltraSSD_LRS": "",
```
This is hard to read. `UltraSSD_LRS` even if commented out should not be within the indexing operator.

rna-afk replied:
gofmt formatted it that way when I tried to create a new line and comment it. I think I'll move it to the top of this line.

rna-afk commented on pkg/asset/machines/master.go:
```diff
@@ -263,6 +263,11 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 		}
 		pool.Platform.Azure = &mpool
 
+		if mpool.OSDisk.DiskType == "Standard_LRS" {
```
This validation is only for the master. The worker will still take up the value. Since the value for the master alone cannot be HDD but it can be for the workers and since there is only one diskType for both master and worker, I thought placing it here would minimize code.

patrickdillon replied:
See https://github.com/openshift/installer/pull/3520#discussion_r417025533

The code in this file is for creating machine assets. The code in the link above is where we validate install configs

rna-afk commented on pkg/types/azure/machinepool.go:
```diff
@@ -38,4 +44,18 @@ func (a *MachinePool) Set(required *MachinePool) {
 		a.OSDisk.DiskSizeGB = required.OSDisk.DiskSizeGB
 	}
+
+	a.OSDisk.DiskType = validateDiskType(required.OSDisk.DiskType)
```
If we were to error out if invalid value is given then I think this code should be in here since it's azure specific.
https://github.com/openshift/installer/blob/6f9a0b8721d3cadfca938e9f4abc9bba3ac292ac/pkg/types/azure/validation/machinepool.go#L14

Making the changes.

patrickdillon commented:
you also will need to add a field for diskType: https://github.com/openshift/installer/blob/master/pkg/types/azure/machinepool.go#L19

rna-afk commented:
With the new commit, I have kind of fixed the top level Machine Pool validation that the master workers in Azure must not use the Standard HDD disks but the top level Platform where the defaultMachinePlatform value is not fixed specifically for the master machines.

abhinavdahiya commented on pkg/types/azure/validation/machinepool.go:
```diff
@@ -12,5 +12,14 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
 	}
+
+	if p.OSDisk.DiskType != "" {
+		//"UltraSSD_LRS": "",
+		_, found := map[string]string{"Standard_LRS": "", "StandardSSD_LRS": "", "Premium_LRS": ""}[p.OSDisk.DiskType]
```
nit:
keep a list of known valid values and then run `sets.NewString(knownDiskTypes...).Has(p.OSDisk.DiskType)`

that helps with readability as a constant list is easy to track and maintain outside the function.

abhinavdahiya commented on pkg/types/azure/validation/machinepool.go:
```diff
@@ -12,5 +12,14 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
+		//"UltraSSD_LRS": "",
+		_, found := map[string]string{"Standard_LRS": "", "StandardSSD_LRS": "", "Premium_LRS": ""}[p.OSDisk.DiskType]
+		if !found {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("diskType"), p.OSDisk.DiskType, "Storage DiskType value invalid"))
```
include the list of supported values in the message.

Also i think there is https://www.godoc.org/k8s.io/apimachinery/pkg/util/validation/field#NotSupported

abhinavdahiya commented on pkg/types/validation/installconfig.go:
```diff
@@ -307,6 +307,9 @@ func validateControlPlane(platform *types.Platform, pool *types.MachinePool, fld
 	if pool.Replicas != nil && *pool.Replicas == 0 {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("replicas"), pool.Replicas, "number of control plane replicas must be positive"))
 	}
+	if pool.Platform.Azure != nil && pool.Name == masterPoolName && pool.Platform.Azure.OSDisk.DiskType == "Standard_LRS" {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("diskType"), pool.Platform.Azure.OSDisk.DiskType, "DiskType not compatible with master machines."))
+	}
```
i think this is redundant, as you already have https://github.com/openshift/installer/pull/3520/files#diff-e8044151bf8d8c8e84c04c49911da440R64-R66

abhinavdahiya commented on pkg/types/validation/installconfig.go:
```diff
@@ -355,6 +358,9 @@ func validatePlatform(platform *types.Platform, fldPath *field.Path, openStackVa
 			return azurevalidation.ValidatePlatform(platform.Azure, c.Publish, f)
 		})
+		if platform.Azure.DefaultMachinePlatform != nil && platform.Azure.DefaultMachinePlatform.OSDisk.DiskType == "Standard_LRS" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child(azure.Name), platform.Azure.DefaultMachinePlatform.OSDisk.DiskType, "Disk type \"Standard_LRS\" not recommended as default disk for machines."))
```
nit, don't use `.`
also try not to use `\"`, you can use ``` `string with "quotes" here` ```

abhinavdahiya commented on pkg/types/validation/machinepools.go:
```diff
@@ -61,6 +61,9 @@ func ValidateMachinePool(platform *types.Platform, p *types.MachinePool, fldPath
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("architecture"), p.Architecture, validArchitectureValues))
 	}
+	if p.Platform.Azure != nil && p.Name == masterPoolName && p.Platform.Azure.OSDisk.DiskType == "Standard_LRS" {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("diskType"), p.Platform.Azure.OSDisk.DiskType, "DiskType not compatible with master machines."))
```
nit :  don;'t have the fieldName in the error message too, it's already part of `fldPath.Child("diskType")`

something like `Standard_LRS is not compatible with control plane machines`

patrickdillon replied:
`master machines` -> `control plane`
Since this is validating the default machine pool perhaps the error message could point the user that this is a valid selection for the control plane machine pool

abhinavdahiya commented on docs/user/azure/customization.md:
```diff
@@ -65,6 +66,7 @@ controlPlane:
       type: Standard_DS4_v2
       osDisk:
         diskSizeGB: 512
+        diskType: Standard_LRS
```
this is not a valid setting with your validations :P

abhinavdahiya commented:
since it's still in WIP.
1. I think we have to allow users to set `UltraSSD_LRS`
2. I think you can add an API validation in `pkg/asset/installconfig/azure/validation.go`
such that you can use the https://godoc.org/github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-10-01/compute#ResourceSkusClient.List
to check if that disk type is available in the location where the cluster is being used.

That would be very helpful imo.

abhinavdahiya commented:
> I requested some changes. I think the biggest issue is we don't know how to handle defaultMachinePool. Master machines cannot have Standard_LRS. So would the ideal way to handle validating the defaultMachinePool be to not allow Standard_LRS for the default pool unless a 
control plane machine pool was provided?

i think it is okay to validate where the information is available completely, if the definition of check is platform specific we can start with keeping it in general function, and later move it to platform specific location when there is more need.

for now, we could also just move all the check for control-plane disk to asset/machines/master.go to keep this simpler.

rna-afk commented:
Fixed most of it (I think). Added validation checks for UltraSSD support but could not create a disk with it because the support for it in Azure terraform is available in version 2.0. There are some deprecation issues that need to be solved if terraform is to be updated. An example of deprecation is

```
Warning: "resource_group_name": [DEPRECATED] This field has been deprecated and is no longer used - will be removed in 2.0 of the Azure Provider (during normal cluster creation)
```

rna-afk commented:
/test golint
/test e2e-aws-upgrade

rna-afk commented:
Will wait for #3526 to complete and then work on the changes for the Ultra SSD.

patrickdillon commented:
If I understood or discussion from scrum yesterday correctly, the plan is to remove Ultra SSD from this PR so it can merge and create a second Ultra SSD PR that depends on #3526.

rna-afk commented:
>If I understood or discussion from scrum yesterday correctly, the plan is to remove Ultra SSD from this PR so it can merge and create a second Ultra SSD PR that depends on #3526.

Yes that's true. I had written the code to check for UltraSSD in a separate function and I have commented the call to the function out. I have also removed the UltraSSD from the list of acceptable disks for now. Let me know if the function that I created needs to be removed.

abhinavdahiya commented on pkg/types/azure/machinepool.go:
```diff
@@ -19,6 +19,8 @@ type MachinePool struct {
 type OSDisk struct {
 	// DiskSizeGB defines the size of disk in GB.
 	DiskSizeGB int32 `json:"diskSizeGB"`
+	// DiskType	defines the type of disk.
```
```suggestion
	// DiskType defines the type of disk.
	// The valid values are <list>
	// By default, <> will be used.
	//
	// +optional
```

abhinavdahiya commented on pkg/types/azure/validation/machinepool.go:
```diff
@@ -12,5 +19,66 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
+	return allErrs
+}
+
+// ValidateRegionForUltraDisks checks that the Ultra SSD disks are available for the user.
+func ValidateRegionForUltraDisks(fldPath *field.Path, region string) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer cancel()
+	client, err := azconfig.NewClient(ctx)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
+	}
+
+	contxt, canc := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer canc()
+
+	_, err = client.GetResourceSku(contxt, region, "UltraSSD_LRS")
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
+	}
+
```
this function depends on pkg/asset, please move it there. pkg/types shuoldn't depend on that package,

abhinavdahiya commented on pkg/types/azure/validation/machinepool.go:
```diff
@@ -12,5 +19,66 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("diskType"), p.Platform.Azure.OSDisk.DiskType, fmt.Sprintf("%s not compatible with control planes.", p.Platform.Azure.OSDisk.DiskType)))
+	}
+
+	if p.Platform.Azure.OSDisk.DiskType != "" {
+		diskTypes := sets.NewString("Standard_LRS", "StandardSSD_LRS", "Premium_LRS") // "UltraSSD_LRS"
+
+		if !diskTypes.Has(p.Platform.Azure.OSDisk.DiskType) {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("diskType"), p.Platform.Azure.OSDisk.DiskType, diskTypes.List()))
+		}
+	}
```
this should be part of `ValidateMachinePool`

abhinavdahiya commented on pkg/types/azure/validation/machinepool.go:
```diff
@@ -12,5 +19,66 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
+		}
+	}
+
+	// if p.OSDisk.DiskType == "UltraSSD_LRS" {
+	// 	allErrs = append(allErrs, ValidateRegionForUltraDisks(fldPath, p)...)
+	// }
```
don't keep commented code.

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +154,29 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
 }
+
+// GetResourceSku checks if the diskType exists for the given region.
+func (c *Client) GetResourceSku(ctx context.Context, region string, diskType string) (*azsku.ResourceSku, error) {
```
the function name and the work don't match. I think we should use a more appropriate name

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +154,29 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+		return nil, err
+	}
+
+	for _, page := range resourceSkus.Values() {
+		if *page.Name == diskType {
+			for _, diskRegion := range *page.Locations {
+				if diskRegion == region {
+					return &page, nil
+				}
+			}
+			return nil, errors.New("ultraSSD_LRS is not available in specified region")
+		}
+	}
```
1. this is only listing one page. there might be a lot more.

abhinavdahiya replied:
1. use autorest/to for ptr to string like changes.

abhinavdahiya replied:
> azsku "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-10-01/compute"

are you sure the compute list skus returns disk SKUs ?? my `az vm list-skus -otable` is returning only virtual machine types skus..

rna-afk replied:
>are you sure the compute list skus returns disk SKUs ?? my az vm list-skus -otable is returning only virtual machine types skus..

I think that the ResourceSkuClient calls [this endpoint](https://docs.microsoft.com/en-us/rest/api/compute/resourceskus/list) to get the list of Resource SKUs available for the subscription. Not sure if they are only virtual machine types. While testing, I did get a list of disks but did not verify if they belong to only Virtual Machine types skus.

patrickdillon replied:
It looks like you can use https://godoc.org/github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-10-01/compute#ResourceSkusResultPage.NextWithContext to advance the pages

rna-afk replied:
ah just got the values in page and multiple pages reference. Making the changes.

rna-afk commented:
/test e2e-azure

patrickdillon commented on pkg/asset/installconfig/azure/mock/azureclient_generated.go:
```diff
@@ -105,8 +106,23 @@ func (m *MockAPI) GetResourcesProvider(ctx context.Context, resourceProviderName
 }
 
-// GetResourcesProvider indicates an expected call of GetResourcesProvider.
+// GetResourcesProvider indicates an expected call of GetResourcesProvider
```
I know this is automated, but it is strange that the periods are being removed from these GoDoc strings. Perhaps this is generated with an outdated version of gomock?

rna-afk replied:
Just installed latest gomock (1.4.3) and tried it. It still removed the period.

patrickdillon replied:
I ran `$ sudo ./hack/go-genmock.sh ` against your PR and it is replacing the periods. `hack/go-genmock.sh` is the canonical way of generating mocks for the installer, so we don't want to commit anything that would deviate from that.

Can you try generating the mocks with `hack/go-genmock.sh` and let me know if you have problems?

patrickdillon commented on pkg/asset/installconfig/azure/mock/azureclient_generated.go:
```diff
@@ -105,8 +106,23 @@ func (m *MockAPI) GetResourcesProvider(ctx context.Context, resourceProviderName
 }
+
+// ValidateResourceSkuAvailability mocks base method
+func (m *MockAPI) ValidateResourceSkuAvailability(ctx context.Context, region, diskType string) (*compute.ResourceSku, error) {
```
Are there any tests using these?

rna-afk replied:
Not yet since these functions are not being called. Will add dummy test cases for now and plan to clean up later

patrickdillon commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -116,3 +117,25 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
 	return field.ErrorList{field.Invalid(fieldPath, p.Region, fmt.Sprintf("region %q does not support resource creation", p.Region))}
 }
+
+// ValidateRegionForUltraDisks checks that the Ultra SSD disks are available for the user.
```
nit: should be lowercase `V` -> `v` to match function name

patrickdillon commented on pkg/types/azure/validation/machinepool.go:
```diff
@@ -12,5 +16,40 @@ func ValidateMachinePool(p *azure.MachinePool, fldPath *field.Path) field.ErrorL
 	}
+
+	if p.OSDisk.DiskType != "" {
+		diskTypes := sets.NewString("Standard_LRS", "StandardSSD_LRS", "Premium_LRS") // "UltraSSD_LRS"
```
nit: Personally I would rather see one of two options:
1. a more substantial comment: e.g. "Adding UltraSSD_LRS depends on..."
2. moving this comment to be part of the data: 
```go
diskTypes := sets.NewString("Standard_LRS", 
                                              // "UltraSSD_LRS", // needs azure terraform provider xx
                                               "StandardSSD_LRS", 
                                               "Premium_LRS") 
``` 
this is my personal preference and not sure if others share the same opinion.
note: I consider this as "data" and am ok leaving in commented data. we should not leave in commented code.

patrickdillon commented on pkg/types/azure/validation/platform.go:
```diff
@@ -20,6 +20,10 @@ func ValidatePlatform(p *azure.Platform, publish types.PublishingStrategy, fldPa
 	if p.DefaultMachinePlatform != nil {
 		allErrs = append(allErrs, ValidateMachinePool(p.DefaultMachinePlatform, fldPath.Child("defaultMachinePlatform"))...)
+		allErrs = append(allErrs, ValidateDefaultDiskType(p.DefaultMachinePlatform, fldPath.Child("defaultMachinePlatform"))...)
+		// if p.DefaultMachinePlatform.OSDisk.DiskType == "UltraSSD_LRS" {
```
commented code related to ultra ssds should be moved to a separate PR, which could be opened immediately.

rna-afk replied:
I keep searching for commented code and this keeps eluding me. Thanks for pointing it out.

patrickdillon commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +155,28 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+		}
+	}
+
+	return nil, errors.New("ultraSSD_LRS not available for specified subscription")
```
It would be better to use `diskType` var in error message. Perhaps `region` as well

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +156,31 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+		return nil, err
+	}
+
+	for !resourceSkus.NotDone() {
+		for _, page := range resourceSkus.Values() {
+			if to.String(page.Name) == diskType {
+				for _, diskRegion := range to.StringSlice(page.Locations) {
+					if diskRegion == region {
+						return &page, nil
+					}
+				}
+			}
+		}
+		resourceSkus.Next()
+	}
```
i would like to keep the code for paging azure apis consistent, like
https://github.com/openshift/installer/blob/2672f9a16e260c1134476c83a384e0d030185f26/pkg/asset/machines/azure/zones.go#L44
or
https://github.com/openshift/installer/blob/2672f9a16e260c1134476c83a384e0d030185f26/pkg/asset/installconfig/azure/dns.go#L148
```suggestion
	for skuPage, err := client.List(ctx); skuPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
		for _, page := range skuPage.Values() {
			if to.String(page.Name) == diskType {
				for _, diskRegion := range to.StringSlice(page.Locations) {
					if strings.EqualFold(diskRegion, region) {
						return &page, nil
					}
				}
			}
		}
	}
```

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +156,31 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
 }
+
+// ValidateResourceSkuAvailability checks if the diskType exists for the given region.
+func (c *Client) ValidateResourceSkuAvailability(ctx context.Context, region string, diskType string) (*azsku.ResourceSku, error) {
```
don't return the sku since this is a validation function, should return only error.

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +156,31 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+		resourceSkus.Next()
+	}
+
+	return nil, errors.New(fmt.Sprintf("diskType ultraSSD_LRS not available for specified subscription in region %s", region))
```
the error should not have `ultraSSD_LRS` hard coded.

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -20,6 +23,7 @@ type API interface {
 	GetControlPlaneSubnet(ctx context.Context, resourceGroupName, virtualNetwork, subnet string) (*aznetwork.Subnet, error)
 	ListLocations(ctx context.Context) (*[]azsubs.Location, error)
 	GetResourcesProvider(ctx context.Context, resourceProviderNamespace string) (*azres.Provider, error)
+	ValidateResourceSkuAvailability(ctx context.Context, region, diskType string) (*azsku.ResourceSku, error)
```
this function feels out of place in the API interface, the rest are getters, while this one is validator.

abhinavdahiya replied:
maybe  convert this to GetDiskSkus(ctx, region)
and then check in the validate function using the list.

abhinavdahiya commented on pkg/asset/installconfig/azure/validation_test.go:
```diff
@@ -162,6 +178,16 @@ func TestAzureInstallConfigValidation(t *testing.T) {
 		},
+		{
+			name:     "Invalid disk types",
+			edits:    editFunctions{removeSubnets, removeVirtualNetwork, invalidateDiskType},
```
nit: why the removeSubnets and removeVirtualNetwork edits ?

abhinavdahiya commented on pkg/types/azure/machinepool.go:
```diff
@@ -27,6 +27,11 @@ type OSDisk struct {
 	//
 	// +kubebuilder:validation:Minimum=0
 	DiskSizeGB int32 `json:"diskSizeGB"`
+	// DiskType defines the type of disk.
+	// The valid values are <list>
+	// By default, <> will be used.
+	// +optional
```
hmm.. this documentation is incomplete.

rna-afk commented:
/test images

patrickdillon commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +157,29 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+
+	var regionPages []azsku.ResourceSku
+
+	for skuPage, err := client.List(ctx); err != nil && skuPage.NotDone(); err = skuPage.NextWithContext(ctx) {
```
It seems like the error is not being handled. If `err != nil` we need to return from `GetDiskSkus`

patrickdillon commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,47 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+	defer cancel()
+	client, err := NewClient(ctx)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
```
if there is an error here should we continue? I think returning here could help you simplify your code later.

patrickdillon commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,47 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+		}
+	}
+
+	return allErrs
```
I think you can simplify.

patrickdillon commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,47 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+			for _, location := range to.StringSlice(page.Locations) {
+				if location == diskType {
+					flag = false
+					break
```
instead of using `flag` can't you just return allErrs here? this should help simplify.

patrickdillon commented on pkg/asset/installconfig/azure/validation_test.go:
```diff
@@ -50,6 +56,12 @@ var (
 	removeVirtualNetwork         = func(ic *types.InstallConfig) { ic.Azure.VirtualNetwork = "" }
 	removeSubnets                = func(ic *types.InstallConfig) { ic.Azure.ComputeSubnet, ic.Azure.ControlPlaneSubnet = "", "" }
+	invalidateDiskType           = func(ic *types.InstallConfig) {
+		ic.Azure.DefaultMachinePlatform.OSDisk.DiskType, ic.Azure.Region = invalidResourceSkuDiskType, validRegion
```
`validInstallConfig` should already have `validRegion` so you shouldn't need to do this. Would be nice to keep formatting consistent and on one line.

patrickdillon commented on pkg/asset/installconfig/azure/validation_test.go:
```diff
@@ -162,6 +178,16 @@ func TestAzureInstallConfigValidation(t *testing.T) {
+		{
+			name:     "Invalid disk types",
+			edits:    editFunctions{invalidateDiskType},
+			errorMsg: "",
```
If there's an invalid disk type you would expect an errorMsg... so if these tests aren't failing it indicates a problem.

rna-afk replied:
These functions are not being called as they are specific to UltraSSD logic. So I left the error msg as empty.

patrickdillon replied:
> Right now, this validate function is not being called. I added these test cases for future purposes.

Does the validate function work if we include it? It looked general to me (like it validates all disk types). or does it only work for Ultra SSD?

rna-afk replied:
It can work in general but usually Azure keeps all types of non-preview disks in all regions, making this check for them unnecessary(I think.. they have let me down before). I made it as general as possible for the future but right now the only use case where we would need the check would be for the UltraSSD disks which is commented.

patrickdillon commented on pkg/asset/installconfig/azure/validation_test.go:
```diff
@@ -50,6 +56,12 @@ var (
+		ic.Azure.DefaultMachinePlatform.OSDisk.DiskType, ic.Azure.Region = invalidResourceSkuDiskType, validRegion
+	}
+	invalidateRegionForDiskType = func(ic *types.InstallConfig) {
+		ic.Azure.DefaultMachinePlatform.OSDisk.DiskType, ic.Azure.Region = validResourceSkuDiskType, invalidResourceSkuRegion
```
I would suggest a separate edit function to add a valid Disk sku.

rna-afk commented:
I actually just put Abhinav's suggestions as is. Was trying to resolve the merge commit and squash it. I will look into the code now and squash later.

rna-afk commented:
/test images

rna-afk commented:
/test images

rna-afk commented:
/test e2e-azure

rna-afk commented:
/test e2e-aws-upgrade

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +157,32 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+
+	for skuPage, err := client.List(ctx); skuPage.NotDone(); err = skuPage.NextWithContext(ctx) {
+		if err != nil {
+			return nil, errors.New(fmt.Sprintf("Error fetching pages from Azure : %s", err.Error()))
```
https://godoc.org/github.com/pkg/errors#Errorf

don't use uppercase to start an error, it doesn't look good when wrapped.

abhinavdahiya replied:
also this could easily just be `errors.Wrap(err, "fetching resource SKUs")`

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +157,32 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+		return regionPages, nil
+	}
+
+	return nil, errors.New(fmt.Sprintf("no disks for specified subscription in region %s", region))
```
https://godoc.org/github.com/pkg/errors#Errorf

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +157,32 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
 	return client, nil
 }
+
+// GetDiskSkus checks if the diskType exists for the given region.
```
the comment is stale.

abhinavdahiya commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -152,3 +157,32 @@ func (c *Client) getProvidersClient(ctx context.Context) (azres.ProvidersClient,
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	var regionPages []azsku.ResourceSku
```
```suggestion
	var skus []azsku.ResourceSku
```

abhinavdahiya commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,37 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+	defer cancel()
+	client, err := NewClient(ctx)
+	if err != nil {
+		return append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
```
```suggestion
		return append(allErrs, field.InternalError(fldPath.Child("diskType"), err))
```

abhinavdahiya commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,37 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+		return append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
+	}
+
+	contxt, canc := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer canc()
```
maybe just set one context for the entire function.

abhinavdahiya commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,37 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+
+	regionDisks, err := client.GetDiskSkus(contxt, region)
+	if err != nil {
+		return append(allErrs, field.InternalError(fldPath.Child("disktype"), err))
```
```suggestion
		return append(allErrs, field.InternalError(fldPath.Child("diskType"), err))
```

abhinavdahiya commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,37 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
 
+// validateRegionForUltraDisks checks that the Ultra SSD disks are available for the user.
+func validateRegionForUltraDisks(fldPath *field.Path, region string) field.ErrorList {
+	allErrs := field.ErrorList{}
```
this is a single field validation function correct? so there is going to be only one error possible. so no need to use ErrorList

simply return `field.InternalError` or `field.Invalid` ..

abhinavdahiya commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -119,6 +120,37 @@ func validateRegion(client API, fieldPath *field.Path, p *aztypes.Platform) fiel
+		}
+	}
+
+	return append(allErrs, field.NotFound(fldPath.Child("disktype"), fmt.Sprintf("%s not found in specified subscription for region", region)))
```
```suggestion
	return append(allErrs, field.NotFound(fldPath.Child("diskType"), fmt.Sprintf("%s is not available in specified subscription for %s", diskType, region)))
```

abhinavdahiya commented on pkg/types/azure/machinepool.go:
```diff
@@ -27,6 +27,13 @@ type OSDisk struct {
 	// +kubebuilder:validation:Minimum=0
 	DiskSizeGB int32 `json:"diskSizeGB"`
+	// DiskType defines the type of disk.
+	// The valid values are <Standard_LRS, Premium_LRS, StandardSSD_LRS>
```
no need to keep the list in `<>`.

abhinavdahiya commented on pkg/types/azure/machinepool.go:
```diff
@@ -27,6 +27,13 @@ type OSDisk struct {
 	DiskSizeGB int32 `json:"diskSizeGB"`
+	// DiskType defines the type of disk.
+	// The valid values are Standard_LRS, Premium_LRS, StandardSSD_LRS
+	// For master nodes, the valid values are Premium_LRS and StandardSSD_LRS.
```
```suggestion
	// For control plane nodes, the valid values are Premium_LRS and StandardSSD_LRS.
```

abhinavdahiya commented:
/approve

abhinavdahiya commented:
this is missing https://github.com/openshift/installer/pull/3595/files#diff-c2be8487621063db19a181fa83200aa3 like updates to customization.md @rna-afk

abhinavdahiya commented:
/test e2e-azure

abhinavdahiya commented:
/retest

abhinavdahiya commented:
/lgtm

/retest
